File: solutions/python/complex-numbers/2/complex_numbers.py
"""Complex numbers and basic complex arithmetic."""
from math import exp, cos, sin


class ComplexNumber:
    """
    Class implementing complex numbers along with ==, +, -, *, /, |z|, z*, exp(z).
    """
    def __init__(self, real, imaginary):
        self.real, self.imaginary = real, imaginary

    def __eq__(self, other):
        if isinstance(other, ComplexNumber):
            return self.real == other.real and self.imaginary == other.imaginary
        return self.imaginary == 0

    def __add__(self, other):
        if isinstance(other, ComplexNumber):
            return ComplexNumber(self.real + other.real,
                                 self.imaginary + other.imaginary)
        return ComplexNumber(self.real + other, self.imaginary)

    def __radd__(self, other):
        return ComplexNumber(self.real + other, self.imaginary)

    def __mul__(self, other):        
        if isinstance(other, ComplexNumber):
            return ComplexNumber(self.real * other.real -
                                 self.imaginary * other.imaginary,
                                 self.real * other.imaginary +
                                 self.imaginary * other.real)
        return ComplexNumber(self.real * other, self.imaginary * other)

    def __rmul__(self, other):
        return ComplexNumber(self.real * other, self.imaginary * other)

    def __sub__(self, other):        
        if isinstance(other, ComplexNumber):
            return ComplexNumber(self.real - other.real,
                                 self.imaginary - other.imaginary)
        return ComplexNumber(self.real - other, self.imaginary)

    def __rsub__(self, other):
        return ComplexNumber(other - self.real, -self.imaginary)

    def __truediv__(self, other):
        if isinstance(other, ComplexNumber):
            return ComplexNumber(
                (self.real * other.real + self.imaginary * other.imaginary) /
                (other.real ** 2 + other.imaginary ** 2),
                (self.imaginary * other.real - self.real * other.imaginary) /
                (other.real ** 2 + other.imaginary ** 2))
        return ComplexNumber(self.real / other, self.imaginary / other)

    def __rtruediv__(self, other):
        return ComplexNumber(self.real * other /
                             (self.real ** 2 + self.imaginary ** 2),
                             -self.imaginary * other /
                             (self.real ** 2 + self.imaginary ** 2))
    
    def __abs__(self):
        return (self.real ** 2 + self.imaginary ** 2) ** 0.5

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        return ComplexNumber(exp(self.real) * cos(self.imaginary),
                             exp(self.real) * sin(self.imaginary))

File: solutions/ruby/complex-numbers/1/complex_numbers.rb
class ComplexNumber
  attr_accessor :real, :imaginary, :abs

  def initialize(real, imaginary = 0.0)
    @real = real
    @imaginary = imaginary
    @abs_sq = real ** 2.0 + imaginary ** 2.0
    @abs = @abs_sq ** 0.5
  end

  def conjugate
    ComplexNumber.new(@real, -@imaginary)
  end

  def exp
    abs_exp = Math::exp(@real)
    cos_im = Math::cos(@imaginary)
    sin_im = (((@imaginary / Math::PI).floor % 2 == 0 ? 1 : -1) * (1.0 - cos_im ** 2.0) ** 0.5)
    ComplexNumber.new(abs_exp * cos_im, abs_exp * sin_im)
  end

  def reciprocal
    ComplexNumber.new(@real / @abs_sq, -(@imaginary / @abs_sq))
  end

  def +(other_num)
    re = (other_num.is_a?(ComplexNumber) ? other_num.real : other_num)
    im = (other_num.is_a?(ComplexNumber) ? other_num.imaginary : 0.0)
    ComplexNumber.new(@real + re, @imaginary + im)
  end

  def -(other_num)
    re = (other_num.is_a?(ComplexNumber) ? other_num.real : other_num)
    im = (other_num.is_a?(ComplexNumber) ? other_num.imaginary : 0.0)
    ComplexNumber.new(@real - re, @imaginary - im)
  end

  def /(other_num)
    return ComplexNumber.new(@real / other_num, (@imaginary / other_num)) unless other_num.is_a?(ComplexNumber)
    return (@imaginary == 0.0 ? mul(self, other_num.reciprocal) : ComplexNumber.new(
      (@real * other_num.real + @imaginary * other_num.imaginary) / other_num.abs ** 2,
      (@imaginary * other_num.real - @real * other_num.imaginary) / other_num.abs ** 2
    ))
  end

  def *(other_num)
    mul(self, other_num)
  end

  def ==(other_num)
    return (@real == other_num && @imaginary == 0.0) unless other_num.is_a?(ComplexNumber)
    #return (@real == other_num.real && @imaginary == other_num.imaginary)
    im_diff = @imaginary - other_num.imaginary
    return (@real == other_num.real && im_diff > -1e-15 && im_diff < 1e-15)
  end

  begin  # private
    private
    
    def mul(c1, c2)
      return ComplexNumber.new(c1.real * c2, c1.imaginary * c2) unless c2.is_a?(ComplexNumber)
      ComplexNumber.new(c1.real * c2.real - c1.imaginary * c2.imaginary, c1.imaginary * c2.real + c1.real * c2.imaginary)
    end
  end  
end
